File: src/game.py
"""Connect Four game engine using numpy arrays for board representation."""

from __future__ import annotations

import numpy as np

ROWS = 6
COLS = 7
WIN_LENGTH = 4


class ConnectFour:
    """Connect Four game state.

    Players are 1 and -1. Player 1 moves first.
    Board uses numpy array: 0=empty, 1=player1, -1=player2.
    """

    __slots__ = ("board", "current_player", "last_move", "_heights", "_move_count")

    def __init__(self) -> None:
        self.board = np.zeros((ROWS, COLS), dtype=np.int8)
        self.current_player: int = 1
        self.last_move: int | None = None
        # Track the next free row for each column (0 = bottom row)
        self._heights = np.zeros(COLS, dtype=np.int8)
        self._move_count: int = 0

    def copy(self) -> ConnectFour:
        """Return a deep copy of this game state."""
        g = ConnectFour.__new__(ConnectFour)
        g.board = self.board.copy()
        g.current_player = self.current_player
        g.last_move = self.last_move
        g._heights = self._heights.copy()
        g._move_count = self._move_count
        return g

    def get_valid_moves(self) -> np.ndarray:
        """Return boolean array of valid columns."""
        return self._heights < ROWS

    def get_valid_move_indices(self) -> list[int]:
        """Return list of valid column indices."""
        return [c for c in range(COLS) if self._heights[c] < ROWS]

    def make_move(self, col: int) -> ConnectFour:
        """Make a move and return self (for chaining). Mutates in-place."""
        row = self._heights[col]
        assert row < ROWS, f"Column {col} is full"
        self.board[row, col] = self.current_player
        self._heights[col] = row + 1
        self.last_move = col
        self._move_count += 1
        self.current_player = -self.current_player
        return self

    def is_win(self) -> bool:
        """Check if the last move created a win."""
        if self.last_move is None:
            return False
        col = self.last_move
        row = self._heights[col] - 1
        player = self.board[row, col]  # The player who just moved
        return self._check_win_at(row, col, player)

    def _check_win_at(self, row: int, col: int, player: int) -> bool:
        """Check if there's a 4-in-a-row through (row, col) for player."""
        directions = [(0, 1), (1, 0), (1, 1), (1, -1)]  # horiz, vert, diag, anti-diag
        for dr, dc in directions:
            count = 1
            # Check positive direction
            for i in range(1, WIN_LENGTH):
                r, c = row + dr * i, col + dc * i
                if 0 <= r < ROWS and 0 <= c < COLS and self.board[r, c] == player:
                    count += 1
                else:
                    break
            # Check negative direction
            for i in range(1, WIN_LENGTH):
                r, c = row - dr * i, col - dc * i
                if 0 <= r < ROWS and 0 <= c < COLS and self.board[r, c] == player:
                    count += 1
                else:
                    break
            if count >= WIN_LENGTH:
                return True
        return False

    def is_draw(self) -> bool:
        """Check if the board is full (draw)."""
        return self._move_count >= ROWS * COLS

    def is_terminal(self) -> bool:
        """Check if the game is over."""
        return self.is_win() or self.is_draw()

    def get_result(self) -> float | None:
        """Get game result from the perspective of the current player.

        Returns:
            1.0 if current player won (shouldn't happen since we switched),
            -1.0 if opponent (whoever just moved) won,
            0.0 for draw,
            None if game is not over.
        """
        if self.is_win():
            # The player who just moved won, so current player lost
            return -1.0
        if self.is_draw():
            return 0.0
        return None

    def get_canonical_board(self) -> np.ndarray:
        """Return board from current player's perspective.

        Current player's pieces are +1, opponent's are -1.
        """
        return self.board * self.current_player

    def encode(self) -> np.ndarray:
        """Encode board as 3-plane tensor for neural network input.

        Plane 0: current player's pieces (1 where current player has a piece)
        Plane 1: opponent's pieces (1 where opponent has a piece)
        Plane 2: constant plane indicating current player (all 1s if player 1, all 0s if player -1)
        """
        canonical = self.get_canonical_board()
        planes = np.zeros((3, ROWS, COLS), dtype=np.float32)
        planes[0] = (canonical == 1).astype(np.float32)
        planes[1] = (canonical == -1).astype(np.float32)
        if self.current_player == 1:
            planes[2] = 1.0
        return planes

    def mirror(self) -> ConnectFour:
        """Return a horizontally mirrored copy of the game state."""
        g = ConnectFour.__new__(ConnectFour)
        g.board = self.board[:, ::-1].copy()
        g.current_player = self.current_player
        g.last_move = COLS - 1 - self.last_move if self.last_move is not None else None
        g._heights = self._heights[::-1].copy()
        g._move_count = self._move_count
        return g

    def display(self, show_col_numbers: bool = True, last_move_highlight: bool = True) -> str:
        """Return a string representation of the board for console display.

        Uses ANSI colors: Red (●) for player 1, Yellow (●) for player -1.
        Last move is highlighted with a red background.
        Grid style:
            1   2   3   4   5   6   7
          +---+---+---+---+---+---+---+
          | · | · | · | · | · | · | · |
          +---+---+---+---+---+---+---+
        """
        RED = "\033[91m"
        YELLOW = "\033[93m"
        RED_BG = "\033[41m"
        RESET = "\033[0m"
        BOLD = "\033[1m"
        DIM = "\033[2m"

        lines: list[str] = []
        separator = "  " + "+---" * COLS + "+"

        # Column numbers header (centered over each cell)
        if show_col_numbers:
            header = "  "
            for c in range(COLS):
                header += f"  {BOLD}{c + 1}{RESET} "
            lines.append(header)

        # Top border
        lines.append(separator)

        # Board rows (top to bottom visually = high row index to low)
        for row in range(ROWS - 1, -1, -1):
            row_str = "  "
            for col in range(COLS):
                piece = self.board[row, col]
                is_last = (
                    last_move_highlight
                    and self.last_move == col
                    and self._heights[col] - 1 == row
                )

                if piece == 1:
                    symbol = f"{RED}●{RESET}"
                elif piece == -1:
                    symbol = f"{YELLOW}●{RESET}"
                else:
                    symbol = f"{DIM}·{RESET}"

                if is_last and piece != 0:
                    row_str += f"|{RED_BG} {symbol}{RED_BG} {RESET}"
                else:
                    row_str += f"| {symbol} "

            row_str += "|"
            lines.append(row_str)
            lines.append(separator)

        return "\n".join(lines)

    def __repr__(self) -> str:
        return f"ConnectFour(move_count={self._move_count}, player={self.current_player})"

File: rust/src/game.rs
/// Connect Four game engine — high-performance bitboard implementation.
///
/// Uses two u64 bitboards (one per player) for O(1) win detection.
/// Board layout: each column uses 7 bits (6 rows + 1 sentinel), left to right.
/// Bit index = col * (ROWS+1) + row, where row 0 = bottom.
///
/// Win detection uses the standard bitboard alignment check:
/// for each direction, shift and AND to find 4 consecutive pieces.
///
/// Players are 1 and -1. Player 1 moves first.

pub const ROWS: usize = 6;
pub const COLS: usize = 7;
pub const WIN_LENGTH: usize = 4;
pub const BOARD_SIZE: usize = ROWS * COLS;

/// Bits per column in the bitboard (ROWS + 1 sentinel bit).
const H1: usize = ROWS + 1;

#[derive(Clone)]
pub struct ConnectFour {
    /// Bitboard for player 1's pieces.
    bb_p1: u64,
    /// Bitboard for player -1's pieces.
    bb_p2: u64,
    pub heights: [u8; COLS],
    pub current_player: i8,
    pub last_move: Option<u8>,
    pub move_count: u8,
}

impl ConnectFour {
    pub fn new() -> Self {
        ConnectFour {
            bb_p1: 0,
            bb_p2: 0,
            heights: [0u8; COLS],
            current_player: 1,
            last_move: None,
            move_count: 0,
        }
    }

    /// Convert (row, col) to bit index in the bitboard.
    #[inline(always)]
    fn bit_index(row: usize, col: usize) -> u32 {
        (col * H1 + row) as u32
    }

    /// Get the value at (row, col): 1, -1, or 0 (empty).
    #[inline]
    pub fn get(&self, row: usize, col: usize) -> i8 {
        let bit = 1u64 << Self::bit_index(row, col);
        if self.bb_p1 & bit != 0 {
            1
        } else if self.bb_p2 & bit != 0 {
            -1
        } else {
            0
        }
    }

    /// Returns a boolean array indicating which columns are not full.
    pub fn get_valid_moves(&self) -> [bool; COLS] {
        let mut valid = [false; COLS];
        for c in 0..COLS {
            valid[c] = (self.heights[c] as usize) < ROWS;
        }
        valid
    }

    /// Returns list of valid column indices.
    pub fn get_valid_move_indices(&self) -> Vec<u8> {
        (0..COLS as u8)
            .filter(|&c| (self.heights[c as usize] as usize) < ROWS)
            .collect()
    }

    /// Make a move in the given column. Mutates in-place.
    /// Panics if the column is full.
    pub fn make_move(&mut self, col: u8) {
        let c = col as usize;
        let row = self.heights[c] as usize;
        assert!(row < ROWS, "Column {} is full", col);

        let bit = 1u64 << Self::bit_index(row, c);
        if self.current_player == 1 {
            self.bb_p1 |= bit;
        } else {
            self.bb_p2 |= bit;
        }

        self.heights[c] += 1;
        self.last_move = Some(col);
        self.move_count += 1;
        self.current_player = -self.current_player;
    }

    /// Check if the given bitboard has a 4-in-a-row.
    ///
    /// For each direction, we shift the bitboard and AND repeatedly.
    /// If any bit remains after 3 shifts, there's a 4-in-a-row.
    #[inline]
    fn has_won(bb: u64) -> bool {
        // Direction offsets for the bitboard layout:
        // Vertical:       shift by 1 (adjacent rows in same column)
        // Horizontal:     shift by H1 (adjacent columns)
        // Diagonal /:     shift by H1+1
        // Anti-diagonal \: shift by H1-1
        const DIRECTIONS: [u32; 4] = [1, H1 as u32, (H1 + 1) as u32, (H1 - 1) as u32];

        for &dir in &DIRECTIONS {
            let m = bb & (bb >> dir);
            if m & (m >> (2 * dir)) != 0 {
                return true;
            }
        }
        false
    }

    /// Check if the last move created a win.
    #[inline]
    pub fn is_win(&self) -> bool {
        if self.last_move.is_none() {
            return false;
        }
        // The player who just moved is -current_player
        let bb = if self.current_player == -1 {
            self.bb_p1
        } else {
            self.bb_p2
        };
        Self::has_won(bb)
    }

    /// Check if the board is full (draw).
    #[inline]
    pub fn is_draw(&self) -> bool {
        self.move_count as usize >= BOARD_SIZE
    }

    /// Check if the game is over.
    pub fn is_terminal(&self) -> bool {
        self.is_win() || self.is_draw()
    }

    /// Get game result from the perspective of the current player.
    ///
    /// Returns:
    ///   Some(-1.0) if opponent (whoever just moved) won,
    ///   Some(0.0) for draw,
    ///   None if game is not over.
    pub fn get_result(&self) -> Option<f64> {
        if self.is_win() {
            // The player who just moved won, so current player lost
            Some(-1.0)
        } else if self.is_draw() {
            Some(0.0)
        } else {
            None
        }
    }

    /// Encode board as 3-plane tensor for neural network input.
    ///
    /// Plane 0: current player's pieces (1 where current player has a piece)
    /// Plane 1: opponent's pieces (1 where opponent has a piece)
    /// Plane 2: constant plane indicating current player (all 1s if player 1, all 0s if player -1)
    ///
    /// Returns flattened [3 * ROWS * COLS] f32 array in C-contiguous order (row-major).
    pub fn encode(&self) -> [f32; 3 * BOARD_SIZE] {
        let mut planes = [0.0f32; 3 * BOARD_SIZE];
        let cp = self.current_player;

        let (bb_cur, bb_opp) = if cp == 1 {
            (self.bb_p1, self.bb_p2)
        } else {
            (self.bb_p2, self.bb_p1)
        };

        for r in 0..ROWS {
            for c in 0..COLS {
                let bit = 1u64 << Self::bit_index(r, c);
                let flat_idx = r * COLS + c;
                // Plane 0: current player's pieces
                if bb_cur & bit != 0 {
                    planes[flat_idx] = 1.0;
                }
                // Plane 1: opponent's pieces
                if bb_opp & bit != 0 {
                    planes[BOARD_SIZE + flat_idx] = 1.0;
                }
            }
        }

        // Plane 2: current player indicator
        if cp == 1 {
            for i in 0..BOARD_SIZE {
                planes[2 * BOARD_SIZE + i] = 1.0;
            }
        }

        planes
    }
}

impl Default for ConnectFour {
    fn default() -> Self {
        Self::new()
    }
}

impl std::fmt::Debug for ConnectFour {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "ConnectFour(move_count={}, player={})",
            self.move_count, self.current_player
        )
    }
}

// ── Tests ──────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new_game() {
        let game = ConnectFour::new();
        assert_eq!(game.current_player, 1);
        assert_eq!(game.move_count, 0);
        assert_eq!(game.last_move, None);
        assert!(!game.is_terminal());
        assert!(game.get_result().is_none());
        // All columns should be valid
        assert!(game.get_valid_moves().iter().all(|&v| v));
    }

    #[test]
    fn test_make_move() {
        let mut game = ConnectFour::new();
        game.make_move(3); // Player 1 in center column
        assert_eq!(game.current_player, -1);
        assert_eq!(game.move_count, 1);
        assert_eq!(game.last_move, Some(3));
        assert_eq!(game.get(0, 3), 1); // Bottom row, center column
        assert_eq!(game.heights[3], 1);
    }

    #[test]
    fn test_stacking_moves() {
        let mut game = ConnectFour::new();
        // Stack pieces in column 0
        for i in 0..ROWS {
            assert!(!game.is_terminal());
            game.make_move(0);
            assert_eq!(game.heights[0] as usize, i + 1);
        }
        // Column 0 should now be full
        let valid = game.get_valid_moves();
        assert!(!valid[0]);
        assert!(valid[1]); // Other columns still valid
    }

    #[test]
    #[should_panic(expected = "Column 0 is full")]
    fn test_full_column_panics() {
        let mut game = ConnectFour::new();
        for _ in 0..ROWS + 1 {
            game.make_move(0);
        }
    }

    #[test]
    fn test_horizontal_win() {
        let mut game = ConnectFour::new();
        // Player 1: columns 0,1,2,3 (bottom row)
        // Player -1: columns 0,1,2 (second row)
        game.make_move(0); // P1
        game.make_move(0); // P-1
        game.make_move(1); // P1
        game.make_move(1); // P-1
        game.make_move(2); // P1
        game.make_move(2); // P-1
        game.make_move(3); // P1 wins horizontally
        assert!(game.is_win());
        assert!(game.is_terminal());
        assert_eq!(game.get_result(), Some(-1.0)); // Current player (P-1) lost
    }

    #[test]
    fn test_vertical_win() {
        let mut game = ConnectFour::new();
        // Player 1 stacks in column 0, Player -1 in column 1
        game.make_move(0); // P1
        game.make_move(1); // P-1
        game.make_move(0); // P1
        game.make_move(1); // P-1
        game.make_move(0); // P1
        game.make_move(1); // P-1
        game.make_move(0); // P1 wins vertically
        assert!(game.is_win());
        assert_eq!(game.get_result(), Some(-1.0));
    }

    #[test]
    fn test_diagonal_win() {
        let mut game = ConnectFour::new();
        // Build a diagonal win for Player 1
        // P1=1, P2=-1
        game.make_move(0); // P1 (0,0)
        game.make_move(1); // P2 (0,1)
        game.make_move(1); // P1 (1,1)
        game.make_move(2); // P2 (0,2)
        game.make_move(3); // P1 (0,3)
        game.make_move(2); // P2 (1,2)
        game.make_move(2); // P1 (2,2)
        game.make_move(3); // P2 (1,3)
        game.make_move(3); // P1 (2,3)
        game.make_move(4); // P2 (0,4)
        game.make_move(3); // P1 (3,3) — diagonal 0,0 -> 1,1 -> 2,2 -> 3,3
        assert!(game.is_win());
    }

    #[test]
    fn test_antidiagonal_win() {
        let mut game = ConnectFour::new();
        // Build anti-diagonal: (0,3), (1,2), (2,1), (3,0)
        // P1=1, P2=-1
        game.make_move(3); // P1 (0,3)
        game.make_move(2); // P2 (0,2)
        game.make_move(2); // P1 (1,2)
        game.make_move(1); // P2 (0,1)
        game.make_move(0); // P1 (0,0)
        game.make_move(1); // P2 (1,1)
        game.make_move(1); // P1 (2,1)
        game.make_move(0); // P2 (1,0)
        game.make_move(0); // P1 (2,0)
        game.make_move(4); // P2 (0,4)
        game.make_move(0); // P1 (3,0) — anti-diagonal (0,3)->(1,2)->(2,1)->(3,0)
        assert!(game.is_win());
    }

    #[test]
    fn test_draw() {
        let mut game = ConnectFour::new();
        // Fill board without winning (known draw pattern)
        let moves = [
            // Col 0: P1,P2,P1,P2,P1,P2
            0, 1, 0, 1, 0, 1,
            // Col 2: P1,P2,P1,P2,P1,P2
            2, 3, 2, 3, 2, 3,
            // Col 4: P1,P2,P1,P2,P1,P2
            4, 5, 4, 5, 4, 5,
            // Col 1: P1,P2,P1,P2,P1,P2
            1, 0, 1, 0, 1, 0,
            // Col 3: P1,P2,P1,P2,P1,P2
            3, 2, 3, 2, 3, 2,
            // Col 5: P1,P2,P1,P2,P1,P2
            5, 4, 5, 4, 5, 4,
            // Col 6: P1,P2,P1,P2,P1,P2
            6, 6, 6, 6, 6, 6,
        ];

        for &col in &moves {
            if game.is_terminal() {
                break;
            }
            game.make_move(col);
        }

        // If we got a draw, great. If not, the pattern caused a win
        // — this test mainly verifies we can fill the board.
        if game.move_count as usize == BOARD_SIZE && !game.is_win() {
            assert!(game.is_draw());
            assert_eq!(game.get_result(), Some(0.0));
        }
    }

    #[test]
    fn test_no_premature_win() {
        let mut game = ConnectFour::new();
        game.make_move(0); // P1
        game.make_move(1); // P2
        game.make_move(2); // P1
        assert!(!game.is_win());
        assert!(!game.is_terminal());
    }

    #[test]
    fn test_clone() {
        let mut game = ConnectFour::new();
        game.make_move(3);
        game.make_move(4);

        let cloned = game.clone();
        assert_eq!(cloned.heights, game.heights);
        assert_eq!(cloned.current_player, game.current_player);
        assert_eq!(cloned.move_count, game.move_count);

        // Mutating original shouldn't affect clone
        game.make_move(5);
        assert_ne!(game.move_count, cloned.move_count);
    }

    #[test]
    fn test_encode_empty_board() {
        let game = ConnectFour::new();
        let encoded = game.encode();

        // Plane 0 (current player = P1): all zeros (no pieces)
        for i in 0..BOARD_SIZE {
            assert_eq!(encoded[i], 0.0);
        }
        // Plane 1 (opponent = P2): all zeros
        for i in 0..BOARD_SIZE {
            assert_eq!(encoded[BOARD_SIZE + i], 0.0);
        }
        // Plane 2 (player indicator): all 1s (player 1)
        for i in 0..BOARD_SIZE {
            assert_eq!(encoded[2 * BOARD_SIZE + i], 1.0);
        }
    }

    #[test]
    fn test_encode_after_moves() {
        let mut game = ConnectFour::new();
        game.make_move(3); // P1 at (0,3)
        game.make_move(4); // P2 at (0,4), now current_player = P1

        let encoded = game.encode();
        // Current player is P1 again
        // Plane 0 (P1 pieces): (0,3) should be 1
        assert_eq!(encoded[0 * COLS + 3], 1.0);
        // Plane 1 (P2 pieces): (0,4) should be 1
        assert_eq!(encoded[BOARD_SIZE + 0 * COLS + 4], 1.0);
        // Plane 2 should be all 1s (current player is P1)
        assert_eq!(encoded[2 * BOARD_SIZE], 1.0);
    }

    #[test]
    fn test_encode_player2_perspective() {
        let mut game = ConnectFour::new();
        game.make_move(3); // P1 at (0,3), now current = P2

        let encoded = game.encode();
        // Current player is P2 (-1)
        // Plane 0 (current player P2's pieces): none yet
        assert_eq!(encoded[0 * COLS + 3], 0.0);
        // Plane 1 (opponent P1's pieces): (0,3) should be 1
        assert_eq!(encoded[BOARD_SIZE + 0 * COLS + 3], 1.0);
        // Plane 2 should be all 0s (current player is -1)
        assert_eq!(encoded[2 * BOARD_SIZE], 0.0);
    }

    #[test]
    fn test_valid_moves_mask() {
        let mut game = ConnectFour::new();
        // Fill column 0
        for _ in 0..ROWS {
            game.make_move(0);
            if game.is_terminal() {
                return; // Can't continue if someone won
            }
            game.make_move(1);
            if game.is_terminal() {
                return;
            }
        }
        let valid = game.get_valid_moves();
        assert!(!valid[0]); // Column 0 should be full
    }

    #[test]
    fn test_get_valid_move_indices() {
        let game = ConnectFour::new();
        let indices = game.get_valid_move_indices();
        assert_eq!(indices, vec![0, 1, 2, 3, 4, 5, 6]);
    }

    #[test]
    fn test_result_is_from_current_player_perspective() {
        let mut game = ConnectFour::new();
        // P1 wins vertically in col 0
        game.make_move(0); // P1
        game.make_move(1); // P2
        game.make_move(0); // P1
        game.make_move(1); // P2
        game.make_move(0); // P1
        game.make_move(1); // P2
        game.make_move(0); // P1 wins

        // After P1 wins, current_player is P2 (-1)
        assert_eq!(game.current_player, -1);
        // Result from P2's perspective: P2 lost → -1.0
        assert_eq!(game.get_result(), Some(-1.0));
    }

    #[test]
    fn test_bitboard_win_detection_speed() {
        // Stress test: play many games and verify win detection is consistent
        let mut game = ConnectFour::new();
        // Horizontal win in bottom row
        game.make_move(0);
        game.make_move(0);
        game.make_move(1);
        game.make_move(1);
        game.make_move(2);
        game.make_move(2);
        assert!(!game.is_win());
        game.make_move(3); // P1 4-in-a-row: cols 0,1,2,3
        assert!(game.is_win());
    }

    #[test]
    fn test_get_value_consistency() {
        let mut game = ConnectFour::new();
        game.make_move(3); // P1 at (0,3)
        assert_eq!(game.get(0, 3), 1);
        assert_eq!(game.get(0, 0), 0);

        game.make_move(4); // P-1 at (0,4)
        assert_eq!(game.get(0, 4), -1);
    }
}
